File: src/format/javadoc/formatter.test.ts
// Unit tests for the javadoc formatter (port of google-java-format's javadoc
// package). Expected outputs were captured from the real gjf jar.

import { test } from "node:test";
import { expect } from "expect";

import { formatJavadoc } from "./formatter.ts";

test("collapses a multi-line comment that fits to one line", () => {
  expect(formatJavadoc("/**\n* foo\n*   bar\n*/", 2)).toBe("/** foo bar */");
});

test("collapses an empty javadoc to /** */", () => {
  expect(formatJavadoc("/**\n */", 2)).toBe("/** */");
});

test("leaves a fitting one-liner unchanged", () => {
  expect(formatJavadoc("/** Tests for foos. */", 0)).toBe("/** Tests for foos. */");
});

test("keeps footer tags one per line at the continuation indent", () => {
  expect(formatJavadoc("/**\n * @param x the x\n * @return y\n */", 2)).toBe(
    "/**\n   * @param x the x\n   * @return y\n   */",
  );
});

test("re-wraps prose to the column limit", () => {
  const input =
    "/**\n   * The subclass should implement {@link #bar()}, with the implementation returning a new instance\n   * of the foo relevant to that baz.\n   */";
  expect(formatJavadoc(input, 2)).toBe(
    "/**\n" +
      "   * The subclass should implement {@link #bar()}, with the implementation returning a new instance\n" +
      "   * of the foo relevant to that baz.\n" +
      "   */",
  );
});

test("a bare tag stays multi-line", () => {
  expect(formatJavadoc("/** @deprecated gone */", 2)).toBe("/**\n   * @deprecated gone\n   */");
});

test("returns input unchanged on a lex failure (unbalanced tag)", () => {
  const input = "/** <pre> unterminated */";
  // Should not throw; gjf returns the input on LexException.
  expect(typeof formatJavadoc(input, 2)).toBe("string");
});

File: togo/internal/format/javadoc/formatter_test.go
package javadoc

import "testing"

func TestFormatJavadoc(t *testing.T) {
	cases := []struct {
		name        string
		input       string
		blockIndent int
		want        string
	}{
		{"collapse to one line", "/**\n* foo\n*   bar\n*/", 2, "/** foo bar */"},
		{"empty collapses", "/**\n */", 2, "/** */"},
		{"fitting one-liner unchanged", "/** Tests for foos. */", 0, "/** Tests for foos. */"},
		{
			"footer tags one per line",
			"/**\n * @param x the x\n * @return y\n */", 2,
			"/**\n   * @param x the x\n   * @return y\n   */",
		},
		{"bare tag stays multi-line", "/** @deprecated gone */", 2, "/**\n   * @deprecated gone\n   */"},
	}
	for _, c := range cases {
		t.Run(c.name, func(t *testing.T) {
			if got := FormatJavadoc(c.input, c.blockIndent); got != c.want {
				t.Errorf("FormatJavadoc(%q, %d):\n got %q\nwant %q", c.input, c.blockIndent, got, c.want)
			}
		})
	}
}

func TestFormatJavadocReturnsInputOnLexFailure(t *testing.T) {
	// Unbalanced <pre> -> LexException -> input returned unchanged.
	in := "/** <pre> unterminated */"
	if FormatJavadoc(in, 2) == "" {
		t.Errorf("expected a string result")
	}
}
